src/compiler/checker.ts

```diff
@@ -15275,7 +15284,11 @@ namespace ts {
                         const inferredType = inferTypeForHomomorphicMappedType(source, target, <IndexType>constraintType);
                         if (inferredType) {
                             const savePriority = priority;
-                            priority |= InferencePriority.HomomorphicMappedType;
+                            // We assign a lower priority to inferences made from types containing non-inferrable
+                            // types because we may only have a partial result (i.e. we may have failed to make
+                            // reverse inferences for some properties).
+                            priority |= getObjectFlags(source) & ObjectFlags.NonInferrableType ?
+                                InferencePriority.PartialHomomorphicMappedType : InferencePriority.HomomorphicMappedType;

```

Comment:
I [experimented with a top-level priority instead of this inner one](https://github.com/weswigham/TypeScript/tree/try-new-top-level-priority). There are some error messages I think it makes look nicer (erroring on "`1` is not assignable to `string`" rather than "`(x: string) => string` is not assignable to `(x: string) => number`" seems way nicer), but there's one test where we issue an error we didn't before (but that might just be because our other priorities are a little off and we're generating a type slightly too derived in the second pass). :(

Reply:
As I mentioned yesterday, I don't think that's the right approach anyway. The `CheckMode.SkipContextSensitive` indicates that context sensitive function expressions should be skipped if any are encountered. But what matters is whether context sensitive functions were _actually_ skipped, and that's tracked by the `ObjectFlags.NonInferrableType` flag. An inference made in the first pass is just as good as an inference made in the second pass because we never infer from anything that contains non-inferrable types. Except in the case of reverse mapped types, and that's why we just need the lower priority for those.